working/0678 - partial classes/feature-brainstorm.md

```diff
@@ -0,0 +1,199 @@
+    firstName: json['firstName'] as String,
+    lastName: json['lastName'] as String,
+    dateOfBirth: DateTime.parse(json['dateOfBirth'] as String),
+  );
+}
+
+Map<String, dynamic> _$PersonToJson(Person instance) => <String, dynamic>{
+      'firstName': instance.firstName,
+      'lastName': instance.lastName,
+      'dateOfBirth': instance.dateOfBirth.toIso8601String(),
+    };
+```
+
+### After (possible future)
+
+```dart
+// after: User-generated code
+import 'package:json_annotation/json_annotation.dart';
+
+part 'example.g.dart';
+
+@JsonSerializable(nullable: false)
+class Person {
+  final String firstName;
+  final String lastName;
+  final DateTime dateOfBirth;
+  Person({this.firstName, this.lastName, this.dateOfBirth});
+
+  // No special wiring needed!
+}
+```
+
+```dart
+// after: machine-generated
+part of 'example.dart';
+
+class Person {
+  // Note: instead of a factory constructor, a "normal" constructor could be
+  // generated that simply populates the fields!
+  factory Person.fromJson(Map<String, dynamic> json) =>
+    Person(
+      firstName: json['firstName'] as String,
+      lastName: json['lastName'] as String,
+      dateOfBirth: DateTime.parse(json['dateOfBirth'] as String),
+    );
+
+  Map<String, dynamic> toJson() => <String, dynamic>{
+      'firstName': firstName,
+      'lastName': lastName,
+      'dateOfBirth': dateOfBirth.toIso8601String(),
+    };
+}
+```
+
+## Design ideas
+
+### Legal to define the same type multiple times in the same library
+
+The following would be legal:
+
+```dart
+class Person{}
+class Person{}
+class Person{}
+```
+
+...and would be treated as
+
+```dart
+class Person{}
+```
+
+> Why not restrict it to one-definition-per-file? It's common for code
+generators to create one file with the output from several separate generators.
+
+### Class hierarchy and members are merged across partial classes
+
+The following would be legal:
+
+```dart
+class Person extends A<int> {
+  final int field;
+  Person(this.field);
+}
+
+class Person implements B<String>{}
+```
+
+...and would be treated as
+
+```dart
+class Person extends A<int> implements B<String>{
+  final int field;
+  Person(this.field);
+}
+```
+
+- Things that are currently not allowed (such as extending more than one type)
+  are still illegal. Implementing multiple interfaces, though, is fine –
+  as long as the interfaces are compatible.
+
+### Defining member "stubs"
+
+The following would be legal:
+
+```dart
+// User-defined class
+class Person {
+  // Reusing `external` keyword here. Not sure if we can/want to add another
+  // keyword
+  external int get value;
+}
+
+// Machine-generated class
+class Person {
+  int get value => 42;
+}
+```
+
+...and would be treated as
+
+```dart
+class Person {
+  int get value => 42;
+}
+```
+
+- The "shape" of `Person` is now visible to static analysis tools even before
+  code generation is ran.
+
+### Open questions
+
+* Do we want a `partial` keyword or similar?
```

Comment:
Sure?

Reply:
I'm sure it's a built-in identifier. I *think* it might be useful.
```dart
part class Foo { 
   int theFoo() => 42;
}
```
I think it reads decently.